Source/Utilities/AsyncOperation.swift

```diff
@@ -0,0 +1,97 @@
+//
+//  AsyncOperation.swift
+//  DigiMeSDK
+//
+//  Created on 28/06/2021.
+//  Copyright © 2021 digi.me Limited. All rights reserved.
+//
+
+import Foundation
+
+class AsyncOperation: Operation {
+    private let lockQueue = DispatchQueue(label: "me.digi.sdk.asyncoperation", attributes: .concurrent)
+    
+    override var isAsynchronous: Bool {
+        true
+    }
+
+    private var _isExecuting = false
+    override private(set) var isExecuting: Bool {
+        get {
+            return lockQueue.sync { () -> Bool in
+                return _isExecuting
+            }
+        }
+        set {
+            willChangeValue(forKey: "isExecuting")
+            lockQueue.sync(flags: [.barrier]) {
+                _isExecuting = newValue
+            }
+            didChangeValue(forKey: "isExecuting")
+        }
+    }
+
+    private var _isFinished = false
+    override private(set) var isFinished: Bool {
+        get {
+            return lockQueue.sync { () -> Bool in
+                return _isFinished
+            }
+        }
+        set {
+            willChangeValue(forKey: "isFinished")
+            lockQueue.sync(flags: [.barrier]) {
+                _isFinished = newValue
+            }
+            didChangeValue(forKey: "isFinished")
+        }
+    }
+
+    override func start() {
+        print("Starting")
+        guard !isCancelled else {
+            finish()
+            return
+        }
+
+        isFinished = false
+        isExecuting = true
+        main()
+    }
+
+    // swiftlint:disable unavailable_function
+    override func main() {
+        fatalError("Subclasses must implement `main` without overriding super.")
```

Comment:
We should probably steer away from fatal errors in the context of the SDK, because this will cause the host application to exit.

Reply:
This one is deliberate as this class is acting as an abstract class and enforces subclasses not calling `super`. It will never be called by client code and is mainly for digi.me devs when making use of this class in future.